documents/extension/funding/service.go

```diff
@@ -0,0 +1,178 @@
+package funding
+
+import (
+	"context"
+	"reflect"
+	"strconv"
+	"strings"
+
+	"github.com/centrifuge/go-centrifuge/utils"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+
+	"github.com/centrifuge/go-centrifuge/centerrors"
+	"github.com/centrifuge/go-centrifuge/documents"
+	"github.com/centrifuge/go-centrifuge/errors"
+	clientfundingpb "github.com/centrifuge/go-centrifuge/protobufs/gen/go/funding"
+)
+
+// Service defines specific functions for extension funding
+type Service interface {
+	documents.Service
+	// DeriveFromPayload derives Funding from clientPayload
+	DeriveFromPayload(ctx context.Context, req *clientfundingpb.FundingCreatePayload, identifier []byte) (documents.Model, error)
+
+	// DeriveFundingResponse returns a funding in client format
+	DeriveFundingResponse(model documents.Model, payload *clientfundingpb.FundingCreatePayload) (*clientfundingpb.FundingResponse, error)
+}
+
+// service implements Service and handles all funding related persistence and validations
+type service struct {
+	documents.Service
+	tokenRegFinder func() documents.TokenRegistry
+}
+
+const fundingLabel = "centrifuge_funding"
+const fundingFieldKey = "centrifuge_funding[{IDX}]."
+const fundingIdx = "{IDX}"
+
+// DefaultService returns the default implementation of the service.
+func DefaultService(
+	srv documents.Service,
+	tokenRegFinder func() documents.TokenRegistry,
+) Service {
+	return service{
+		Service:        srv,
+		tokenRegFinder: tokenRegFinder,
+	}
+}
+
+func generateKey(idx, fieldName string) string {
+	return strings.Replace(fundingFieldKey, fundingIdx, idx, -1) + fieldName
+}
+
+func keyFromJSONTag(idx, jsonTag string) (key string, err error) {
+	correctJSONParts := 2
```

Comment:
we cannot restrict the tags to 2. IMO, we dont need to check for this. The first word in the tag is always name of the field

Reply:
It was a workaround for removing the protobuf related fields.

`	XXX_NoUnkeyedLiteral struct{}                 `json:"-"`

Otherwise I need to check if the variable name is not equal to `-`

Reply:
Anyway it changes if I use a internal struct